eodatasets3/properties.py

```diff
@@ -413,10 +415,34 @@ def dataset_version(self) -> str:
         """
         return self.properties.get("odc:dataset_version")
 
+    @property
+    def collection_number(self) -> str:
+        """
+        The version of the collection.
+        Eg:
+          metadata:
+            product_family: wofs
+            dataset_version: 1.6.0
+            collection_number: 3
```

Comment:
This is a weird one, because for our ARD we use the USGS collection number. Not sure what we should actually be using here, should it be `3` (our collection) or `1` (the USGS source collection)?

Perhaps a question for @simonaoliver?

Reply:
From how I understand the naming system, since it's derived from our collection 3 data it should use that as its collection number. Only the Level 1 processed by USGS is using the USGS collection numbers (I believe?).

In the spec, the first number of the `dataset_version` is the collection number, hence ard datasets having `dataset_version: 3.0.0` etc -- and that's how the [current collection_number property reads it](https://github.com/GeoscienceAustralia/eo-datasets/blob/c115ee1542f3a4a43f9d076aaec4a3f021692d8b/eodatasets3/model.py#L182-L185). If we want to copy it in a stand-alone property, it should match, I think? (do we need it separate?)

Reply:
@jeremyh In this case we need separation as the sample agreed upon S3 path as `/ga_ls_wo_3/1-6-0/096/095/2000/10/01/ga_ls_wo_3_096095_2000-10-01_final_water.tiff`

We need `dataset_version` to be `1.6.0` in the path also, `_3`. Alex has different opinion as well, the `_3` part can come from part of product itself.

Reply:
Check this out, @jeremyh : https://data.dea.ga.gov.au/baseline/ga_ls8c_ard_3/089/076/2016/06/22/ga_ls8c_ard_3-0-0_089076_2016-06-22_final.odc-metadata.yaml

`landsat:collection_number: 1`

Reply:
> Check this out, @jeremyh : https://data.dea.ga.gov.au/baseline/ga_ls8c_ard_3/089/076/2016/06/22/ga_ls8c_ard_3-0-0_089076_2016-06-22_final.odc-metadata.yaml
> 
> `landsat:collection_number: 1`

Ah, yeah. That's specifically in the landsat namespace (all fields of which match verbatim with USGS's MTL document field names/values), so it's not a normal general "collection number".

Reply:
> @jeremyh In this case we need separation as the sample agreed upon S3 path as `/ga_ls_wo_3/1-6-0/096/095/2000/10/01/ga_ls_wo_3_096095_2000-10-01_final_water.tiff`
> 
> We need `dataset_version` to be `1.6.0` in the path also, `_3`. Alex has different opinion as well, the `_3` part can come from part of product itself.

So is the second folder a wofs algorithm version or something? I might have missed these discussions, but in previous discussion everyone was very strict on what each component of `dataset_version` was, and eodatasets had to be fixed to match. Might need @simonaoliver to confirm

Reply:
For the record, I don't think the standard is perfect, but the only thing worse than having imperfect standard fields, is having imperfect standard fields that are used inconsistently everywhere. So I have to play the character of nitpicky bureaucrat that demands specs :slightly_smiling_face:

Reply:
The idea is that collection 3's `_3` is included in the product name (explicitly or implicitly, I don't care here!) and is separate to the product version number.

Reply:
To sum:
On the Alchemist side 

1 - dataset versions for `wofs`  1.6.0 and for `fc` 2.5.0. 
2 - collection_number is 3
3 - The derivatives s3 filepath structure is `ga_ls_wo_3/1-6-0/096/095/2000/10/01/ga_ls_wo_3_096095_2000-10-01_final_water.tif`

@jeremyh @alexgleith if we are happy, we can merge

Reply:
I'm happy, so long as 1 and 2 are configurable from Alchemist and not hardcoded! Which I think they are.

Reply:
To re-ask the same question by @jeremyh 

Is the second folder an algorithm version number?

Reply:
Sorry @sixy6e 

The second folder is the product version number, which is also called the `odc:dataset_version` in the ARD yamls.

So, look at CMI: https://cmi.ga.gov.au/data-products/dea/119/dea-fractional-cover-landsat, it's the `2.2.1` number.

Reply:
> In the spec, the first number of the `dataset_version` is the collection number, hence ARD datasets having `dataset_version: 3.0.0` etc -- and that's how the [current collection_number property reads it](https://github.com/GeoscienceAustralia/eo-datasets/blob/c115ee1542f3a4a43f9d076aaec4a3f021692d8b/eodatasets3/model.py#L182-L185). If we want to copy it in a stand-alone property, it should match, I think? (do we need it separate?)

There was a decision in one of Simon's meetings about naming conventions recently where we decoupled product version numbers (what you're calling `dataset_version`) from collection numbers.

So for derived products, we need a new field for the `collection_number`, since it's not the first digit of the product number.

Reply:
But product version and dataset version are different things. They're not substitutable with one another.

If it is desired to use a product version number, there should be a new field added, i.e. `product_version`.  Not use the product version in the namespace for `dataset_version`.  Otherwise not only are we then inconsistent in file naming, but also in the metadata where a single field would then be used to describe different entities.

The reason I ask is that maybe we should change ARD to be consistent with other products based on the new decision.  In which case the ARD product should use a product version instead of a dataset version.

Reply:
And looking at the CMI link, the 2.2.1 number is pointing to Collection-2. The first digit `2` there is referring to the collection.

Reply:
Regarding the first digit of a `product_version` being the collection number, we discussed this in Simon's naming conventions meeting, and agreed to not continue this practice. Plus, there are always exceptions, and NIDEM is 1.0.0 currently https://data.dea.ga.gov.au/?prefix=nidem/

But I am curious as to what the difference is between a `dataset_version` and a `product_version`, @sixy6e?

Reply:
A `product_version` relates to the tracking of changes relating to the state of the product, which could be algorithmic changes.
A `dataset_version` is a specific instance of a product generated according to the specifications of the product.

Updating the `product_version` most of the time would also update the `dataset_version`.  When it wouldn't is when there are major changes in the definition or generation of the product that deviates significantly from the previous specification.  No longer idempotent, or not within the spec tolerance.  In which a new specification for a product definition, meaning a reset on the `dataset_version`.

The `dataset_version` can be updated independently of the product version and allows for increased traceability and provenance.  For example, updating `eo-datasets`, or other libraries, don't change the product, but changes the way the product was generated to still meet the specifications.  In this instance, the `dataset_version` is updated enabling traceability for a specific instance of a product.

If the intent is to be updating the modules that generate the product, and not record the instance, then we've lost a level of provenance.

Eg generating WOfS with an updated ODC version, may not change the product but is now a different instance of that product.  We can use the same ODC but change the WOfS algorithm (ever so slightly to avoid major differences), is an updated product, as well as an instance of the product.